juju/framework.py

```diff
@@ -541,6 +557,25 @@ def snapshot(self):
     def restore(self, snapshot):
         self._cache = snapshot
 
+    def clear_dirty(self):
+        self._dirty = False
+
+    def set_dirty(self):
+        self._dirty = True
+
+    def is_dirty(self):
+        """Check if a copy of a stored state data is dirty in memory."""
+        try:
+            dirty = getattr(self, '_dirty')
+        except AttributeError:
+            # When StoredStateData is restored from persistent storage its __init__ is not called so it is clear that it is not dirty yet.
+            dirty = False
+        return dirty
+
+    def on_pre_commit(self, event):
+        self.framework.save_snapshot(self)
+        self.clear_dirty()
```

Comment:
```python
    def on_pre_commit(self, event):
        if self.is_dirty():
            self.framework.save_snapshot(self)
            self.clear_dirty()
```

Reply:
Per discussion today, the dirty flag checking will be removed and the optimization split into a separate PR.  This will still be relevant there, though, I think.